### PR TITLE
xAPI plugin

### PR DESCRIPTION
**Pull request by AngelFQC**

### CODE REVIEW

**Comment by ywarnier** [comment]
:partying_face:

**Comment by lonesomewalker** [comment]
Is this the REAL xAPI ? :D 
In 1.11.x?
Is there a matching LRS (learning record store), or are we free to choose?

**Comment by AngelFQC** [comment]
> Is this the REAL xAPI ? :D
> In 1.11.x?
> Is there a matching LRS (learning record store), or are we free to choose?
You can choose and configure the LRS in the plugin settings

**Comment by AngelFQC** [comment]
> Is this the REAL xAPI ? :D
> In 1.11.x?
> Is there a matching LRS (learning record store), or are we free to choose?

@lonesomewalker You can choose and configure the LRS in the plugin settings

File: main/inc/lib/hook/HookLearningPathEnd.php [hunk 1]
```diff
@@ -0,0 +1,29 @@
+<?php
+/* For licensing terms, see /license.txt */
```
Comment (ywarnier) [inline main/inc/lib/hook/HookLearningPathEnd.php:2]: Tenía entendido que ahora era necesario poner una línea blanca entre `<?php` y `/* For licensing terms, ...`

File: plugin/xapi/src/Entity/Actor.php [hunk 2]
```diff
@@ -0,0 +1,161 @@
+<?php
+
+/*
+ * This file is part of the xAPI package.
+ *
+ * (c) Christian Flothmann <christian.flothmann@xabbuh.de>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
```
Comment (ywarnier) [inline plugin/xapi/src/Entity/Actor.php:9]: En el caso de código externo insertado en Chamilo, es necesario colocar el archivo de LICENSE indicado en la raíz de la carpeta donde se usa.

**Review by ywarnier — CHANGES_REQUESTED** [review]

**Comment by ywarnier** [comment]
> Is this the REAL xAPI ? :D

Yes

> In 1.11.x?

Yes, it changes the minimum requirement of 1.11.14 from PHP 5.6 to PHP 7.1 though (I doubt that's an issue, but so you know...). It will still be considered beta-level for now though, as we need a lot of testing to ensure this is enough for most people.
Bear in mind that there are 2 "sides" to xAPI: the integration of new tools, LTI-style, and the integration of SCORM-style contents in the learning paths. These are 2 different aspects. To our knowledge, other Open Source LMSes do not implement these two aspects yet (only the LTI-style part).

> Is there a matching LRS (learning record store), or are we free to choose?

Yes, Chamilo implementes its own LRS but you could also point to some external LRS. This might generate some lags/delays for now though, as sending frequent messages to an external LRS is still something that is a bit too ambitious of xAPI in our opinion.

**Comment by lonesomewalker** [comment]
BOY THIS IS LIKE CHRISTMAS FOR ME!

> ... it changes the minimum requirement of 1.11.14 from PHP 5.6 to PHP 7.1 though (I doubt that's an issue, but so you know...).

Well, new things tend to need new dependencies, but 7.0 is already outdated. Most hosting providers offer 7.0 AND later so this shouldn't be an issue for the people which install in webhosting space.

> It will still be considered beta-level for now though, as we need a lot of testing to ensure this is enough for most people.

NEVERMIND, we will test this hard; especially in Germany there are so many potential enterprise users which just want SCORM and xAPI which Chamilo was not capable of in the past.
I know that this doesn't mean that the software gives you advantage in didactical things, but you know...
This is a HUGE step for Chamilo in a sub-release??? :1st_place_medal: 

> ... Chamilo implementes its own LRS but you could also point to some external LRS. This might generate some lags/delays for now though, as sending frequent messages to an external LRS is still something that is a bit too ambitious of xAPI in our opinion.

I am fine for now with the integrated one :-)

**Comment by ywarnier** [comment]
> BOY THIS IS LIKE CHRISTMAS FOR ME!

Merry Christmas, then ;-)

> > It will still be considered beta-level for now though, as we need a lot of testing to ensure this is enough for most people.
> 
> NEVERMIND, we will test this hard; especially in Germany there are so many potential enterprise users which just want SCORM and xAPI which Chamilo was not capable of in the past.
> I know that this doesn't mean that the software gives you advantage in didactical things, but you know...
> This is a HUGE step for Chamilo in a sub-release??? 

This is a plugin, so you need to enable it. We couldn't have made it part of this sub-release as core because it required many database and structure changes, but as a plugin it is allowed to create its own structure.

It is developed to be integrated into core in Chamilo 2.0 (hopefully early next year). Having it as a plugin will allow us to test it in order to have it fully working in 2.0 (which will then be considered a big step forward).

But yes... @AngelFQC has been hard at work on this for a while, so we're very happy this is out just one week before 1.11.14 is released (hopefully).

**Comment by AngelFQC** [comment]
Closing this PR.
A partial implementation was sent in #3680

**Pull request closed without merging**